keps/sig-api-machinery/20190721-explain-categories-expand.md

```diff
@@ -0,0 +1,198 @@
+lists all the fields of a Kind in the alphabetic order, and there is not enough information in the declaration of these fields to list them in a specific order.
+
+We propose to add an annotation `x-kubernetes-explain-category` applicable to fields, that represents both
+the name of the category and the order of this category related to the other categories of the same Kind.
+
+In addition, we propose to add an annotation `x-kubernetes-explain-expand` applicable to fields, 
+that indicates if the field should be expanded in documentation in order to display inline the content 
+of this field.
+
+## Proposal
+
+### User Stories
+
+#### Categorize the fields of PodSpec
+
+To categorize a field, it is necessary to add a comment to the field in `types.go`:
+
+```
+// +k8s:openapi-gen=x-kubernetes-explain-category:1.category-name
+```
+
+The value (here *1.category-name*) contains two parts, separated by a dot (.).
+The first part is an integer that indicates the order of the category related to the
+other categories of the field. The second part is a name of the category, with spaces
+replaced by dashes (-).
+
+Example:
+```go
+type PodSpec struct {
+  // +k8s:openapi-gen=x-kubernetes-explain-category:2.volumes
+  Volumes []Volume
+  // +k8s:openapi-gen=x-kubernetes-explain-category:1.containers
+  InitContainers []Container
+  // +k8s:openapi-gen=x-kubernetes-explain-category:1.containers
+  Containers []Container
+  // +k8s:openapi-gen=x-kubernetes-explain-category:4.lifecycle
+  RestartPolicy RestartPolicy
+  [...]
+```
+
+As a result, the documentation of PodSpec would be written by `kubectl explain` as:
+```
+KIND:     Pod
+VERSION:  v1
+
+RESOURCE: spec <Object>
+
+DESCRIPTION:
+     ...
+     
+FIELDS:
+  CONTAINERS
+
+   initContainers        <[]Object>
+     ...
+
+   containers            <[]Object>
+     ...
+
+  VOLUMES
+
+   volumes               <[]Object>
+     ...
+
+  LIFECYCLE
+  
+   restartPolicy         <Object>
+    ...
+```
+
+#### Expand some fields of Container
+
+To indicate that a field should be expanded in the documentation, it is necessary to add a comment to the field in `types.go`:
+
+```
+// +k8s:openapi-gen=x-kubernetes-explain-expand:
+```
+
+Example:
+
+```go
+type Container struct {
+  // +k8s:openapi-gen=x-kubernetes-explain-expand:
+  Ports []ContainerPort
+  [...]
+}
+```
+
+As a result, the documentation of Container would be written by `kubectl explain` as:
+
+```
+KIND:     Pod
+VERSION:  v1
+
+RESOURCE: containers <[]Object>
+
+DESCRIPTION:
+     ...
+
+FIELDS:
+
+   [first Container fields...]
+
+   ports     <[]Object>
+     <description of ports>
+
+     hostIP     <string>
+       <description of hostIP>
+
+     hostPort   <sint32>
+       <description of hostPort>
+
+     [other ports fields]
+
+   [other Container fields...]
+```
+
+#### Order fields of VolumeMount
+
+In order to sort fields without indicating a category, it is necessary to add a comment to the field in `types.go`:
```

Comment:
Field display order is a separate problem than categorization (in my mind, at least). We could (and should) automatically propagate the order as defined in the source files with some effort; humans deliberately and carefully defined the fields in that order and it's almost criminal to throw that away like we do.

But there's no way to automatically categorize them--that seems useful.